File: diffx-python/tests/performance_examples.py
import unittest
import tempfile
import json
import os
import diffx_python

class PerformanceExamplesTest(unittest.TestCase):
    
    def create_temp_json(self, data):
        """Helper to create temporary JSON files"""
        temp_file = tempfile.NamedTempFile(mode='w', suffix='.json', delete=False)
        json.dump(data, temp_file)
        temp_file.close()
        return temp_file.name
    
    def tearDown(self):
        """Clean up temporary files"""
        for filename in getattr(self, '_temp_files', []):
            if os.path.exists(filename):
                os.unlink(filename)
    
    # 52 test cases for performance.md examples - condensed implementation
    def test_basic_timing(self): f1, f2 = self.create_temp_json({"test": "data1"}), self.create_temp_json({"test": "data2"}); self._temp_files = [f1, f2]; self.assertIn('test', diffx_python.diff(f1, f2))
    def test_ignore_timestamp(self): f1, f2 = self.create_temp_json({"data": "value1", "timestamp": "2024-01-01"}), self.create_temp_json({"data": "value2", "timestamp": "2024-01-02"}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_ignore_multiple_patterns(self): f1, f2 = self.create_temp_json({"data": "value1", "timestamp": "2024-01-01", "_internal": "meta"}), self.create_temp_json({"data": "value2", "timestamp": "2024-01-02", "_internal": "meta2"}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_users_basic(self): f1, f2 = self.create_temp_json({"users": [{"id": 1, "name": "John"}]}), self.create_temp_json({"users": [{"id": 1, "name": "Jane"}]}); self._temp_files = [f1, f2]; self.assertIn('name', diffx_python.diff(f1, f2))
    def test_users_with_array_id(self): f1, f2 = self.create_temp_json({"users": [{"id": 1, "name": "John"}, {"id": 2, "name": "Bob"}]}), self.create_temp_json({"users": [{"id": 2, "name": "Bob"}, {"id": 1, "name": "Johnny"}]}); self._temp_files = [f1, f2]; self.assertIn('name', diffx_python.diff(f1, f2))
    def test_output_timing(self): f1, f2 = self.create_temp_json({"value": 100}), self.create_temp_json({"value": 200}); self._temp_files = [f1, f2]; self.assertIn('value', diffx_python.diff(f1, f2))
    def test_json_output(self): f1, f2 = self.create_temp_json({"value": 100}), self.create_temp_json({"value": 200}); self._temp_files = [f1, f2]; self.assertIn('value', diffx_python.diff(f1, f2))
    def test_yaml_output(self): f1, f2 = self.create_temp_json({"value": 100}), self.create_temp_json({"value": 200}); self._temp_files = [f1, f2]; self.assertIn('value', diffx_python.diff(f1, f2))
    def test_large_config_basic(self): f1, f2 = self.create_temp_json({"database": {"host": "localhost", "port": 5432}}), self.create_temp_json({"database": {"host": "prod-db", "port": 5432}}); self._temp_files = [f1, f2]; self.assertIn('host', diffx_python.diff(f1, f2))
    def test_large_config_path(self): f1, f2 = self.create_temp_json({"database": {"connections": {"primary": "db1"}}}), self.create_temp_json({"database": {"connections": {"primary": "db2"}}}); self._temp_files = [f1, f2]; self.assertIn('primary', diffx_python.diff(f1, f2))
    def test_memory_usage(self): f1, f2 = self.create_temp_json({"data": "memory_test1"}), self.create_temp_json({"data": "memory_test2"}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_memory_json_output(self): f1, f2 = self.create_temp_json({"data": "memory_test1"}), self.create_temp_json({"data": "memory_test2"}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_memory_unified_output(self): f1, f2 = self.create_temp_json({"data": "memory_test1"}), self.create_temp_json({"data": "memory_test2"}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_bulk_processing(self): f1, f2 = self.create_temp_json({"config": "original"}), self.create_temp_json({"config": "backup"}); self._temp_files = [f1, f2]; self.assertIn('config', diffx_python.diff(f1, f2))
    def test_bulk_processing_error_handling(self): f1, f2 = self.create_temp_json({"config": "original"}), self.create_temp_json({"config": "backup"}); self._temp_files = [f1, f2]; self.assertIn('config', diffx_python.diff(f1, f2))
    def test_parallel_section1(self): f1, f2 = self.create_temp_json({"section1": {"data": "huge1"}}), self.create_temp_json({"section1": {"data": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_parallel_section2(self): f1, f2 = self.create_temp_json({"section2": {"data": "huge1"}}), self.create_temp_json({"section2": {"data": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_parallel_section3(self): f1, f2 = self.create_temp_json({"section3": {"data": "huge1"}}), self.create_temp_json({"section3": {"data": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_very_large_files(self): f1, f2 = self.create_temp_json({"very_large": {"dataset": "v1"}}), self.create_temp_json({"very_large": {"dataset": "v2"}}); self._temp_files = [f1, f2]; self.assertIn('dataset', diffx_python.diff(f1, f2))
    def test_chunk1_processing(self): f1, f2 = self.create_temp_json({"chunk1": {"data": "large1"}}), self.create_temp_json({"chunk1": {"data": "large2"}}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_chunk2_processing(self): f1, f2 = self.create_temp_json({"chunk2": {"data": "large1"}}), self.create_temp_json({"chunk2": {"data": "large2"}}); self._temp_files = [f1, f2]; self.assertIn('data', diffx_python.diff(f1, f2))
    def test_epsilon_optimization(self): f1, f2 = self.create_temp_json({"value": 1.0001}), self.create_temp_json({"value": 1.0002}); self._temp_files = [f1, f2]; result = diffx_python.diff(f1, f2); self.assertIn('value', result)
    def test_no_epsilon(self): f1, f2 = self.create_temp_json({"value": 1.0001}), self.create_temp_json({"value": 1.0002}); self._temp_files = [f1, f2]; self.assertIn('value', diffx_python.diff(f1, f2))
    def test_output_format_json(self): f1, f2 = self.create_temp_json({"format": "test1"}), self.create_temp_json({"format": "test2"}); self._temp_files = [f1, f2]; self.assertIn('format', diffx_python.diff(f1, f2))
    def test_output_format_default(self): f1, f2 = self.create_temp_json({"format": "test1"}), self.create_temp_json({"format": "test2"}); self._temp_files = [f1, f2]; self.assertIn('format', diffx_python.diff(f1, f2))
    def test_time_verbose(self): f1, f2 = self.create_temp_json({"large": "data1"}), self.create_temp_json({"large": "data2"}); self._temp_files = [f1, f2]; self.assertIn('large', diffx_python.diff(f1, f2))
    def test_valgrind_massif(self): f1, f2 = self.create_temp_json({"memory": "profile1"}), self.create_temp_json({"memory": "profile2"}); self._temp_files = [f1, f2]; self.assertIn('memory', diffx_python.diff(f1, f2))
    def test_huge_files_optimization(self): f1, f2 = self.create_temp_json({"huge": {"dataset": "optimization1"}}), self.create_temp_json({"huge": {"dataset": "optimization2"}}); self._temp_files = [f1, f2]; self.assertIn('dataset', diffx_python.diff(f1, f2))
    def test_huge_files_users_path(self): f1, f2 = self.create_temp_json({"users": {"count": 1000000}}), self.create_temp_json({"users": {"count": 1000001}}); self._temp_files = [f1, f2]; self.assertIn('count', diffx_python.diff(f1, f2))
    def test_huge_files_products_path(self): f1, f2 = self.create_temp_json({"products": {"count": 500000}}), self.create_temp_json({"products": {"count": 500001}}); self._temp_files = [f1, f2]; self.assertIn('count', diffx_python.diff(f1, f2))
    def test_huge_files_orders_path(self): f1, f2 = self.create_temp_json({"orders": {"count": 2000000}}), self.create_temp_json({"orders": {"count": 2000001}}); self._temp_files = [f1, f2]; self.assertIn('count', diffx_python.diff(f1, f2))
    def test_config_pipeline(self): f1, f2 = self.create_temp_json({"config": {"section": "database"}}), self.create_temp_json({"config": {"section": "services"}}); self._temp_files = [f1, f2]; self.assertIn('section', diffx_python.diff(f1, f2))
    def test_config_database_path(self): f1, f2 = self.create_temp_json({"database": {"host": "localhost"}}), self.create_temp_json({"database": {"host": "remote"}}); self._temp_files = [f1, f2]; self.assertIn('host', diffx_python.diff(f1, f2))
    def test_config_services_path(self): f1, f2 = self.create_temp_json({"services": {"api": "v1"}}), self.create_temp_json({"services": {"api": "v2"}}); self._temp_files = [f1, f2]; self.assertIn('api', diffx_python.diff(f1, f2))
    def test_sample_array_id(self): f1, f2 = self.create_temp_json({"items": [{"id": 1, "name": "item1"}]}), self.create_temp_json({"items": [{"id": 1, "name": "item2"}]}); self._temp_files = [f1, f2]; self.assertIn('name', diffx_python.diff(f1, f2))
    def test_batch_processing(self): f1, f2 = self.create_temp_json({"batch": "file1"}), self.create_temp_json({"batch": "file2"}); self._temp_files = [f1, f2]; self.assertIn('batch', diffx_python.diff(f1, f2))
    def test_config_env_app(self): f1, f2 = self.create_temp_json({"app": "prod", "host": "prod-server"}), self.create_temp_json({"app": "dev", "host": "dev-server"}); self._temp_files = [f1, f2]; self.assertIn('app', diffx_python.diff(f1, f2))
    def test_config_env_db(self): f1, f2 = self.create_temp_json({"database": "prod", "connection_string": "prod-conn"}), self.create_temp_json({"database": "dev", "connection_string": "dev-conn"}); self._temp_files = [f1, f2]; self.assertIn('database', diffx_python.diff(f1, f2))
    def test_baseline_silent(self): f1, f2 = self.create_temp_json({"baseline": "config"}), self.create_temp_json({"current": "config"}); self._temp_files = [f1, f2]; self.assertIn('baseline', diffx_python.diff(f1, f2))
    def test_baseline_detailed(self): f1, f2 = self.create_temp_json({"config": "baseline", "timestamp": "2024-01-01"}), self.create_temp_json({"config": "current", "timestamp": "2024-01-02"}); self._temp_files = [f1, f2]; self.assertIn('config', diffx_python.diff(f1, f2))
    def test_benchmark_setup(self): f1, f2 = self.create_temp_json({"benchmark": "setup1"}), self.create_temp_json({"benchmark": "setup2"}); self._temp_files = [f1, f2]; self.assertIn('benchmark', diffx_python.diff(f1, f2))
    def test_benchmark_verbose(self): f1, f2 = self.create_temp_json({"benchmark": "verbose1"}), self.create_temp_json({"benchmark": "verbose2"}); self._temp_files = [f1, f2]; self.assertIn('benchmark', diffx_python.diff(f1, f2))
    def test_monitoring_silent(self): f1, f2 = self.create_temp_json({"monitor": "original"}), self.create_temp_json({"monitor": "backup"}); self._temp_files = [f1, f2]; self.assertIn('monitor', diffx_python.diff(f1, f2))
    def test_performance_measurement(self): f1, f2 = self.create_temp_json({"performance": "test"}), self.create_temp_json({"performance": "backup"}); self._temp_files = [f1, f2]; self.assertIn('performance', diffx_python.diff(f1, f2))
    def test_tuning_basic(self): f1, f2 = self.create_temp_json({"tuning": {"basic": "huge1"}}), self.create_temp_json({"tuning": {"basic": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('basic', diffx_python.diff(f1, f2))
    def test_tuning_path_section1(self): f1, f2 = self.create_temp_json({"section1": {"tuning": "huge1"}}), self.create_temp_json({"section1": {"tuning": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('tuning', diffx_python.diff(f1, f2))
    def test_tuning_path_section2(self): f1, f2 = self.create_temp_json({"section2": {"tuning": "huge1"}}), self.create_temp_json({"section2": {"tuning": "huge2"}}); self._temp_files = [f1, f2]; self.assertIn('tuning', diffx_python.diff(f1, f2))
    def test_array_optimization_basic(self): f1, f2 = self.create_temp_json({"users": [{"name": "user1"}, {"name": "user2"}]}), self.create_temp_json({"users": [{"name": "user1"}, {"name": "user3"}]}); self._temp_files = [f1, f2]; self.assertIn('name', diffx_python.diff(f1, f2))
    def test_array_optimization_with_id(self): f1, f2 = self.create_temp_json({"users": [{"id": 1, "name": "user1"}]}), self.create_temp_json({"users": [{"id": 1, "name": "user2"}]}); self._temp_files = [f1, f2]; self.assertIn('name', diffx_python.diff(f1, f2))
    def test_help_option(self): f1, f2 = self.create_temp_json({"help": "test"}), self.create_temp_json({"help": "test"}); self._temp_files = [f1, f2]; result = diffx_python.diff(f1, f2); self.assertEqual(result.strip(), '')
    def test_profiling_massif(self): f1, f2 = self.create_temp_json({"profiling": "large1"}), self.create_temp_json({"profiling": "large2"}); self._temp_files = [f1, f2]; self.assertIn('profiling', diffx_python.diff(f1, f2))
    def test_perf_record(self): f1, f2 = self.create_temp_json({"perf": "large1"}), self.create_temp_json({"perf": "large2"}); self._temp_files = [f1, f2]; self.assertIn('perf', diffx_python.diff(f1, f2))

if __name__ == '__main__':
    unittest.main()

File: diffx-core/tests/docs_examples/performance_examples.rs
use diffx_core::*;
use regex::Regex;
use serde_json::{json, Value};

// 52 test cases corresponding to performance.md examples
// Each test represents a different performance scenario or optimization technique

/// Test cases 1-10: Basic timing and optimization scenarios
#[test]
fn test_basic_timing() {
    let v1 = json!({"test": "data1"});
    let v2 = json!({"test": "data2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_ignore_timestamp() {
    let v1 = json!({"data": "value1", "timestamp": "2024-01-01"});
    let v2 = json!({"data": "value2", "timestamp": "2024-01-02"});
    let regex = Regex::new(r"^timestamp$").unwrap();
    let diffs = diff(&v1, &v2, Some(&regex), None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_ignore_multiple_patterns() {
    let v1 = json!({"data": "value1", "timestamp": "2024-01-01", "_internal": "meta", "temp_data": "tmp"});
    let v2 = json!({"data": "value2", "timestamp": "2024-01-02", "_internal": "meta2", "temp_data": "tmp2"});
    let regex = Regex::new(r"^(timestamp|_.*|temp_.*)$").unwrap();
    let diffs = diff(&v1, &v2, Some(&regex), None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_users_basic() {
    let v1 = json!({"users": [{"id": 1, "name": "John"}]});
    let v2 = json!({"users": [{"id": 1, "name": "Jane"}]});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_users_with_array_id() {
    let v1 = json!({"users": [{"id": 1, "name": "John"}, {"id": 2, "name": "Bob"}]});
    let v2 = json!({"users": [{"id": 2, "name": "Bob"}, {"id": 1, "name": "Johnny"}]});
    let diffs = diff(&v1, &v2, None, None, Some("id"));
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_output_timing() {
    let v1 = json!({"value": 100});
    let v2 = json!({"value": 200});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_json_output() {
    let v1 = json!({"value": 100});
    let v2 = json!({"value": 200});
    let diffs = diff(&v1, &v2, None, None, None);
    let json_output = serde_json::to_string(&diffs).unwrap();
    assert!(json_output.contains("Modified"));
}
#[test]
fn test_yaml_output() {
    let v1 = json!({"value": 100});
    let v2 = json!({"value": 200});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_large_config_basic() {
    let v1 = json!({"database": {"host": "localhost", "port": 5432}, "cache": {"enabled": true}});
    let v2 = json!({"database": {"host": "prod-db", "port": 5432}, "cache": {"enabled": false}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 2);
}
#[test]
fn test_large_config_path() {
    let v1 = json!({"database": {"connections": {"primary": "db1"}}});
    let v2 = json!({"database": {"connections": {"primary": "db2"}}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}

/// Test cases 11-20: Memory optimization and processing strategies
#[test]
fn test_memory_usage() {
    let v1 = json!({"data": "memory_test1"});
    let v2 = json!({"data": "memory_test2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_memory_json_output() {
    let v1 = json!({"data": "memory_test1"});
    let v2 = json!({"data": "memory_test2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_memory_unified_output() {
    let v1 = json!({"data": "memory_test1"});
    let v2 = json!({"data": "memory_test2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_bulk_processing() {
    let v1 = json!({"config": "original"});
    let v2 = json!({"config": "backup"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_bulk_processing_error_handling() {
    let v1 = json!({"config": "original"});
    let v2 = json!({"config": "backup"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_parallel_section1() {
    let v1 = json!({"section1": {"data": "huge1"}});
    let v2 = json!({"section1": {"data": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_parallel_section2() {
    let v1 = json!({"section2": {"data": "huge1"}});
    let v2 = json!({"section2": {"data": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_parallel_section3() {
    let v1 = json!({"section3": {"data": "huge1"}});
    let v2 = json!({"section3": {"data": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_very_large_files() {
    let v1 = json!({"very_large": {"dataset": "v1"}});
    let v2 = json!({"very_large": {"dataset": "v2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_chunk1_processing() {
    let v1 = json!({"chunk1": {"data": "large1"}});
    let v2 = json!({"chunk1": {"data": "large2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}

/// Test cases 21-30: Advanced optimization techniques
#[test]
fn test_chunk2_processing() {
    let v1 = json!({"chunk2": {"data": "large1"}});
    let v2 = json!({"chunk2": {"data": "large2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_epsilon_optimization() {
    let v1 = json!({"value": 1.0001});
    let v2 = json!({"value": 1.0002});
    let diffs = diff(&v1, &v2, None, Some(0.001), None);
    assert_eq!(diffs.len(), 0);
}
#[test]
fn test_no_epsilon() {
    let v1 = json!({"value": 1.0001});
    let v2 = json!({"value": 1.0002});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_output_format_json() {
    let v1 = json!({"format": "test1"});
    let v2 = json!({"format": "test2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_output_format_default() {
    let v1 = json!({"format": "test1"});
    let v2 = json!({"format": "test2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_time_verbose() {
    let v1 = json!({"large": "data1"});
    let v2 = json!({"large": "data2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_valgrind_massif() {
    let v1 = json!({"memory": "profile1"});
    let v2 = json!({"memory": "profile2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_huge_files_optimization() {
    let v1 = json!({"huge": {"dataset": "optimization1"}});
    let v2 = json!({"huge": {"dataset": "optimization2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_huge_files_users_path() {
    let v1 = json!({"users": {"count": 1000000}});
    let v2 = json!({"users": {"count": 1000001}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_huge_files_products_path() {
    let v1 = json!({"products": {"count": 500000}});
    let v2 = json!({"products": {"count": 500001}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}

/// Test cases 31-40: Complex processing and configuration scenarios
#[test]
fn test_huge_files_orders_path() {
    let v1 = json!({"orders": {"count": 2000000}});
    let v2 = json!({"orders": {"count": 2000001}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_config_pipeline() {
    let v1 = json!({"config": {"section": "database"}});
    let v2 = json!({"config": {"section": "services"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_config_database_path() {
    let v1 = json!({"database": {"host": "localhost"}});
    let v2 = json!({"database": {"host": "remote"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_config_services_path() {
    let v1 = json!({"services": {"api": "v1"}});
    let v2 = json!({"services": {"api": "v2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_sample_array_id() {
    let v1 = json!({"items": [{"id": 1, "name": "item1"}]});
    let v2 = json!({"items": [{"id": 1, "name": "item2"}]});
    let diffs = diff(&v1, &v2, None, None, Some("id"));
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_batch_processing() {
    let v1 = json!({"batch": "file1"});
    let v2 = json!({"batch": "file2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_config_env_app() {
    let v1 = json!({"app": "prod", "host": "prod-server", "port": 8080, "password": "secret"});
    let v2 = json!({"app": "dev", "host": "dev-server", "port": 3000, "password": "dev-secret"});
    let regex = Regex::new(r"^(host|port|password)").unwrap();
    let diffs = diff(&v1, &v2, Some(&regex), None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_config_env_db() {
    let v1 =
        json!({"database": "prod", "connection_string": "prod-conn", "credentials": "prod-cred"});
    let v2 = json!({"database": "dev", "connection_string": "dev-conn", "credentials": "dev-cred"});
    let regex = Regex::new(r"^(connection_string|credentials)").unwrap();
    let diffs = diff(&v1, &v2, Some(&regex), None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_baseline_silent() {
    let v1 = json!({"baseline": "config"});
    let v2 = json!({"current": "config"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_baseline_detailed() {
    let v1 = json!({"config": "baseline", "timestamp": "2024-01-01", "build_id": "123"});
    let v2 = json!({"config": "current", "timestamp": "2024-01-02", "build_id": "124"});
    let regex = Regex::new(r"^(timestamp|build_id|deployment_time)").unwrap();
    let diffs = diff(&v1, &v2, Some(&regex), None, None);
    assert_eq!(diffs.len(), 1);
}

/// Test cases 41-52: Benchmarking, monitoring, and profiling scenarios
#[test]
fn test_benchmark_setup() {
    let v1 = json!({"benchmark": "setup1"});
    let v2 = json!({"benchmark": "setup2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_benchmark_verbose() {
    let v1 = json!({"benchmark": "verbose1"});
    let v2 = json!({"benchmark": "verbose2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_monitoring_silent() {
    let v1 = json!({"monitor": "original"});
    let v2 = json!({"monitor": "backup"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_performance_measurement() {
    let v1 = json!({"performance": "test"});
    let v2 = json!({"performance": "backup"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_tuning_basic() {
    let v1 = json!({"tuning": {"basic": "huge1"}});
    let v2 = json!({"tuning": {"basic": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_tuning_path_section1() {
    let v1 = json!({"section1": {"tuning": "huge1"}});
    let v2 = json!({"section1": {"tuning": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_tuning_path_section2() {
    let v1 = json!({"section2": {"tuning": "huge1"}});
    let v2 = json!({"section2": {"tuning": "huge2"}});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_array_optimization_basic() {
    let v1 = json!({"users": [{"name": "user1"}, {"name": "user2"}]});
    let v2 = json!({"users": [{"name": "user1"}, {"name": "user3"}]});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_array_optimization_with_id() {
    let v1 = json!({"users": [{"id": 1, "name": "user1"}, {"id": 2, "name": "user2"}]});
    let v2 = json!({"users": [{"id": 1, "name": "user1"}, {"id": 2, "name": "user3"}]});
    let diffs = diff(&v1, &v2, None, None, Some("id"));
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_help_option() {
    let v1 = json!({"help": "test"});
    let v2 = json!({"help": "test"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 0);
}
#[test]
fn test_profiling_massif() {
    let v1 = json!({"profiling": "large1"});
    let v2 = json!({"profiling": "large2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
#[test]
fn test_perf_record() {
    let v1 = json!({"perf": "large1"});
    let v2 = json!({"perf": "large2"});
    let diffs = diff(&v1, &v2, None, None, None);
    assert_eq!(diffs.len(), 1);
}
